paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngWriter.kt

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (C) 2023 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package app.cash.paparazzi.internal.apng
+
+import app.cash.paparazzi.internal.ImageUtils
+import app.cash.paparazzi.internal.ImageUtils.resize
+import app.cash.paparazzi.internal.apng.PngConstants.Header
+import app.cash.paparazzi.internal.apng.PngConstants.Header.ACTL
+import app.cash.paparazzi.internal.apng.PngConstants.Header.FCTL
+import app.cash.paparazzi.internal.apng.PngConstants.Header.FDAT
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IDAT
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IEND
+import app.cash.paparazzi.internal.apng.PngConstants.Header.IHDR
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_BITS_PER_PIXEL
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_COLOR_TYPE_RGB
+import app.cash.paparazzi.internal.apng.PngConstants.PNG_COLOR_TYPE_RGBA
+import okio.Buffer
+import okio.BufferedSink
+import okio.FileSystem
+import okio.Path
+import okio.Path.Companion.toPath
+import okio.buffer
+import java.awt.Rectangle
+import java.awt.image.BufferedImage
+import java.io.Closeable
+import java.io.File
+import java.lang.IllegalStateException
+import java.lang.Integer.max
+import java.util.zip.CRC32
+import java.util.zip.Deflater
+
+/**
+ * Provides a basic implementation for encoding APNG's.
+ *
+ * Refs: https://www.w3.org/TR/png/#4Concepts.Encoding, https://wiki.mozilla.org/APNG_Specification
+ */
+internal class ApngWriter(
+  private val path: Path,
+  private val totalFrameCount: Int,
+  fps: Int,
+  fileSystem: FileSystem = FileSystem.SYSTEM
+) : Closeable {
+
+  private val fpsNum = 1
+  private val fpsDen = fps
+  private val tempFile = File(FileSystem.SYSTEM_TEMPORARY_DIRECTORY.toFile(), "temp-${path.name}").apply { deleteOnExit() }
```

Comment:
Okio is hurting you here more than helping
```suggestion
  private val tempFile = File.createTempFile(path.name, "apng").apply { deleteOnExit() }
```

Reply:
Also can we delete this eagerly in `close()` rather than with `deleteOnExit()`? We shouldn't make assumptions for how long the VM will be alive.

Reply:
Also also, have you considered a peer `Path` with `.tmp` or something suffixed? That would allow you keep all writes going through the supplied `FileSystem` rather than needing to rely on the system one (better again for tests).

Reply:
Re the last from @JakeWharton, can you assume that the input path's parent is writable? Isn't the temp a safer bet to create new files?